### PR TITLE
[WIP] Dom diffing

### PR DESCRIPTION
**Pull request by cromefire**
Much like the image doc checks, but using the Dom.

### CODE REVIEW

**Comment by cromefire** [comment]
@prateekbh is there any specific reason why the builds are on osx?

**Comment by artipost** [comment]
Failed pictures (156): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/5acd65be8ef9c0e116d5040b71b0e3ed/raw/a004af61cff0c65e5d6aa221b109ba30e7e0e9f9/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409130443](https://travis-ci.org/prateekbh/preact-material-components/builds/409130443), for commit [7afaea5](https://github.com/prateekbh/preact-material-components/pull/882/commits/7afaea58352073c18e865332a73ec0bbe00eaaf8).

**Comment by cromefire** [comment]
The problem is that I can't reproduce them completely especially with all the webdriver stuff and I would like to do that.

**Comment by cromefire** [comment]
Generally it's way easier to spin up a Ubuntu vm to fully reproduce everything as it is with getting your hands on a mac.

**Comment by prateekbh** [comment]
I am personally not in favor of dom diff testing due to the following reason:

a.) Its super tough to compare the nitigrities of a large text doc(but i guess a tool can take of this for us).

b.) Bigger problem, the DOM is not constructed by us, a small change which requires no change rom us will start breaking for us. Simply put, we are not constructing the DOM, we're not the owner of it.. so  not sure why should we test it.

**Comment by cromefire** [comment]
I mean I don't think, it's possible to do vdom diffing right?

**Comment by cromefire** [comment]
We could try to serialize the VNode to json

**Comment by prateekbh** [comment]
lets come back to this, once we get the TS PR in

**Comment by cromefire** [comment]
I wanted to use this to ensure the TS PR doesn't break anything, that was the main purpose

**Comment by cromefire** [comment]
Because there are this failures of the pictures and I want to know where this comes from

**Comment by artipost** [comment]
Failed pictures (0): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/d0b84aa82c854c14de880ea191067cc4/raw/15cb0ecb3e219d1701294bfdf0fe3f5cb5d208e7/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409322140](https://travis-ci.org/prateekbh/preact-material-components/builds/409322140), for commit [50ea90c](https://github.com/prateekbh/preact-material-components/pull/882/commits/50ea90c9d99edd23ed26a5a86bbbce7c0308db4f).

**Comment by artipost** [comment]
Failed pictures (0): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/6de41077ac39e9e38cb75a61dd1bcca9/raw/15cb0ecb3e219d1701294bfdf0fe3f5cb5d208e7/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409325736](https://travis-ci.org/prateekbh/preact-material-components/builds/409325736), for commit [23dc44c](https://github.com/prateekbh/preact-material-components/pull/882/commits/23dc44ce1841070a710da4a8de4e7695d7e29277).

File: docs/tests/domdiff.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  for (const driver_desc of drivers) {
+    console.log(`Testing ${page} using ${driver_desc.name}`);
+    const {driver, name} = driver_desc;
+    await driver.get(`http://localhost:8080/${page}`);
+    await driver.sleep(2000);
+    const gen_dom = pretty(await driver.getPageSource(), {
+      ocd: true
+    })
+      .replace(
+        new RegExp('src="/(bundle|polyfills)([.]|-)[.a-zA-Z0-9]+[.]js"'),
```
Comment (cromefire) [inline docs/tests/domdiff.js]: Can someone tell me what's wrong with these regexp's? They work in the Firefox console.

File: docs/tests/golden/dom/firefox/component/card.html [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+                </div>
+              </div>
+            </div>
+          </div>
+        </div>
+      </div>
+    </div>
+    <script defer="defer" src="/bundle.0233d.js"></script>
+    <script>
+      window.fetch || document.write('<script src="/polyfills.b4b0f.js"><\/script>')
```
Comment (cromefire) [inline docs/tests/golden/dom/firefox/component/card.html]: should match this

**Comment by artipost** [comment]
Failed DOMs (0): [failed-doms.zip](https://gist.githubusercontent.com/artipost/a84decbce190f2fed0ed646f2e507d95/raw/14feba73572110c496fb2fef51caba0dedf28129/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409327187](https://travis-ci.org/prateekbh/preact-material-components/builds/409327187), for commit [5bba6ad](https://github.com/prateekbh/preact-material-components/pull/882/commits/5bba6ade1bac1900d5a4b2bc2c5447e49487e079).

File: docs/tests/domdiff.js (hunk 1 above)
Comment (YePpHa) [reply docs/tests/domdiff.js]: You can use https://regex101.com/ to test your regex. It seems like the `/` after `src="` is not escaped.

Comment (cromefire) [reply docs/tests/domdiff.js]: Using `new RegExp("[...]")` instead of `/[...]/`

Comment (cromefire) [reply docs/tests/domdiff.js]: but yes will check that out

**Comment by artipost** [comment]
Failed DOMs (2): [failed-doms.zip](https://gist.githubusercontent.com/artipost/4010750e32e1b1cad12e67d1d4a42a92/raw/e2f2659c17b25a9880d0a2c6df3844a32b26c765/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409536198](https://travis-ci.org/prateekbh/preact-material-components/builds/409536198), for commit [93f9509](https://github.com/prateekbh/preact-material-components/pull/882/commits/93f950986415d35f6734f8d6d8c44d2f3ff7c92c).

**Comment by artipost** [comment]
Failed DOMs (27): [failed-doms.zip](https://gist.githubusercontent.com/artipost/98761270234acfaddd80423502944414/raw/a2b8bb8f7b5b13387fded471a84c5df087ade9bb/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409541400](https://travis-ci.org/prateekbh/preact-material-components/builds/409541400), for commit [0baab28](https://github.com/prateekbh/preact-material-components/pull/882/commits/0baab280ad0e83ccae90db93fe35f72c4eff0b41).

**Comment by artipost** [comment]
Failed DOMs (27): [failed-doms.zip](https://gist.githubusercontent.com/artipost/a9047fc83c561440a26373b9ae6accba/raw/d543cc04a81adc648e2ebb25fa7b47a43cff8689/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409542536](https://travis-ci.org/prateekbh/preact-material-components/builds/409542536), for commit [67ff537](https://github.com/prateekbh/preact-material-components/pull/882/commits/67ff53746f33445005e53dd0beea4662c0f49ada).

**Comment by artipost** [comment]
Failed DOMs (27): [failed-doms.zip](https://gist.githubusercontent.com/artipost/6a6a207ddb5f21c2c07fa60a4c8901ac/raw/7e503ed6a1822721a42569a0d1d12fe7431ea1ee/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409545341](https://travis-ci.org/prateekbh/preact-material-components/builds/409545341), for commit [9bbcc5c](https://github.com/prateekbh/preact-material-components/pull/882/commits/9bbcc5cd814151b325dd35def5ee983907446d24).

**Comment by artipost** [comment]
Failed DOMs (1): [failed-doms.zip](https://gist.githubusercontent.com/artipost/88ad52d4842b66efbc245670a9063ccd/raw/a81ed2fa68f8cd44e556fe76f0311b4d289bfdb6/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409549470](https://travis-ci.org/prateekbh/preact-material-components/builds/409549470), for commit [2facb6e](https://github.com/prateekbh/preact-material-components/pull/882/commits/2facb6e5b06eaa16de7e108e31968b8a2ed1acd4).

**Comment by artipost** [comment]
Failed DOMs (1): [failed-doms.zip](https://gist.githubusercontent.com/artipost/f0a71ee7eb75f98accb346bee82fe7b6/raw/f45317a5cd770b2e0e1352c2b665f9daab1b52d2/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409551423](https://travis-ci.org/prateekbh/preact-material-components/builds/409551423), for commit [c092f2c](https://github.com/prateekbh/preact-material-components/pull/882/commits/c092f2c9207227ea75d379bfe45ebcbf69a65979).

**Comment by artipost** [comment]
Failed DOMs (1): [failed-doms.zip](https://gist.githubusercontent.com/artipost/927edd2dd47b4c072d5aae8719c737cd/raw/700e203bac876cb57b049a474d2714d7b10ca799/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409555044](https://travis-ci.org/prateekbh/preact-material-components/builds/409555044), for commit [be7b293](https://github.com/prateekbh/preact-material-components/pull/882/commits/be7b293c0350b28cb496f1d97e2b74f566b0f1b5).

**Comment by artipost** [comment]
Failed DOMs (52): [failed-doms.zip](https://gist.githubusercontent.com/artipost/c07d448b739938e126aa7b2071e63f48/raw/1bd1867f308a0c17d95392ba2a62e70fbeaf29c5/failed-doms.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [409576546](https://travis-ci.org/prateekbh/preact-material-components/builds/409576546), for commit [27a19e7](https://github.com/prateekbh/preact-material-components/pull/882/commits/27a19e7561f902adcb4bed2dce571090e12d2129).

**Comment by cromefire** [comment]
@prateekbh please have a look again, it's really good for debugging invisible(/hard visible) bugs, I could speed up the image tests also

**Comment by prateekbh** [comment]
I would still prefer something like enzyme to check what we produce instead of the end result because there are minor updates when styles and DOM changes to `mdc-web` not us.

so i wanted a minimal setup to maintain.

I am a little hesitant to add this till a point where there is a more active maintainer support.

**Pull request closed without merging**